control-plane/csi-driver/src/bin/controller/controller.rs

```diff
@@ -178,11 +178,44 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
         tracing::trace!(request = ?args);
         let _permit = self.create_volume_permit().await?;
 
-        if args.volume_content_source.is_some() {
-            return Err(Status::invalid_argument(
-                "Source for create volume is not supported",
-            ));
-        }
+        let volume_content_source = if let Some(source) = args.volume_content_source {
+            match source.r#type {
+                Some(Type::Snapshot(snapshot_source)) => {
+                    let snapshot_uuid =
+                        Uuid::parse_str(&snapshot_source.snapshot_id).map_err(|_e| {
+                            Status::invalid_argument(format!(
+                                "Malformed snapshot UUID: {}",
+                                snapshot_source.snapshot_id
+                            ))
+                        })?;
+
+                    if IoEngineApiClient::get_client()
+                        .get_volumes_snapshot(&snapshot_uuid)
+                        .await?
+                        .state
+                        .ready_as_source
+                    {
+                        Some(snapshot_uuid)
+                    } else {
+                        return Err(Status::failed_precondition(
+                            format!("Volume creation from snapshot source is not allowed for {snapshot_uuid}"),
+                        ));
+                    }
```

Comment:
Shouldn't need to do this, create_volume will validate this

Reply:
Cool. Removing this then.